app/beacon-omop-worker/beacon_omop_worker/query_solvers.py

```diff
@@ -307,3 +309,167 @@ def solve_individuals(db_manager: SyncDBManager, query_terms: str):
         return response_summary
     except Exception as e:
         logger.error(str(e))
+
+class KaplanMeierQuerySolver:
+
+    def __init__(self, db_manager) -> None:
+        self.db_manager = db_manager
+
+    def _get_concept_id(self, vocab: str, concept_code: str) -> str:
+        """
+        Get concept_id from vocabulary and concept code.
+        
+        Args:
+            vocab (str): The vocabulary name (e.g., 'SNOMED').
+            concept_code (str): The concept code (e.g., '386661006').
+        
+        Returns:
+            str: The concept_id corresponding to the vocabulary and concept code.
+        """
+        logger = logging.getLogger(config.LOGGER_NAME)
+        try:
+            query = select(Concept.concept_id).where(
+                and_(
+                    Concept.vocabulary_id == vocab,
+                    Concept.concept_code == concept_code
+                )
+            )
+            result = pd.read_sql(query, con=self.db_manager.engine.connect())
+            concept_id = str(result["concept_id"].values[0])
+            logger.info(f"Found concept_id {concept_id} for {vocab}:{concept_code}.")
+            return concept_id
+        except Exception as e:
+            logger.error(f"Error fetching concept_id for {vocab}:{concept_code}: {e}")
+            raise
+
+    def fetch_condition_data(self, concept_id: str) -> pd.DataFrame:
+        """
+        Fetch data from the OMOP database based on concept_id.
+        
+        Args:
+            concept_id (str): The concept_id for the condition.
+        
+        Returns:
+            pd.DataFrame: DataFrame containing the condition data.
+        """
+        logger = logging.getLogger(config.LOGGER_NAME)
+        query = text(
+            f"""
+            SELECT
+                condition_occurrence.condition_occurrence_id,
+                condition_occurrence.condition_concept_id,
+                concept.concept_name,
+                condition_occurrence.condition_start_date,
+                condition_occurrence.condition_end_date,
+                person.birth_datetime,
+                person.gender_source_value,
+                person.race_source_value,
+                location.city,
+                location.county, 
+                death.death_date,
+                death.cause_concept_id
+            FROM
+                condition_occurrence
+            JOIN
+                concept ON condition_occurrence.condition_concept_id = concept.concept_id
+            JOIN
+                person ON condition_occurrence.person_id = person.person_id
+            JOIN
+                location ON person.location_id = location.location_id
+            LEFT JOIN
+                death ON condition_occurrence.person_id = death.person_id
+            WHERE
+                condition_occurrence.condition_concept_id = :concept_id
+            """
```

Comment:
We've found in previous query types that it is best to load the tables into memory and then perform the joins. The database is simply too slow. What you should do is load the individual tables into pandas, limiting the data from those tables with a `WHERE` clause if possible. Then join the tables using pandas.

Reply:
Modified